design/recovery-internals.md

```diff
@@ -0,0 +1,193 @@
+
+Recall that `ServerDBInfo` has master's interface and is propogated by CC to every process in a cluster. The current running tLogs can use the master interface in its `ServerDBInfo` to send itself's interface to master.
+Master simply waits on receiving the `TLogRejoinRequest` streams: for each tLog’s interface received, the master compares the interface ID with the tLog ID read from cstate. Once the master collects enough old tLog interfaces, it will use the interfaces to lock those tLogs.
+The logic of collecting tLogs’ interfaces is implemented in `trackRejoins()` function.
+The logic of locking the tLogs is implemented in `epochEnd()` function in TagPartitionedLogSystems.actor.cpp.
+
+
+Once we lock the cstate, we bump the recoveryCount and write the cstate to signal to the other recovery attempts that they should not recruit more tLogs. If we do not do this, there could be many recovery attempts that recruit tLogs and cause the system to run out of memory. This operation is the reason why every recovery bumps the recoveryCount (epoch) by 2 instead of 1.
+
+
+*How does each tLog know the current master’s interface?*
+
+Master interface is stored in `serverDBInfo`. Once the CC recruits the master, it updates the `serverDBInfo` with the master’s interface. CC will send the updated `serverDBInfo`, which has the master’s interface, to all processes. tLog processes (i,e., tLog workers) monitor the `serverDBInfo` in an actor. when the `serverDBInfo` changes, it will register itself to the new master. The logic for a tLog worker to monitor `serverDBInfo` change is implemented in `monitorServerDBInfo()` actor.
+
+
+*How does each role, such as tLog and data distributor (DD), register its interface to master and CC?*
+
+* tLog monitors `serverDBInfo` change and sends its interface to the new master; 
+
+* Data distributor (DD) and ratekeeper rejoin themselves to CC because they are no longer a part of the recovery process;
+
+* Storage server (SS) does not rejoin. It waits for the tLogs to be ready and commit their interfaces into database with a special transaction.
+
+
+## Phase 3: RECRUITING
+
+Once the master locks the cstate, it will recruit the still-alive tLogs from the previous generation for the benefit of faster recovery. The master gets the old tLogs’ interfaces from the READING_CSTATE phase and uses those interfaces to track which old tLog are still alive, the implementation of which is in `trackRejoins()`. 
+
+
+Once the master gets enough tLogs, it calculates the knownCommittedVersion, which is the maximum durable version from the still-alive tLogs in the previous generation. The master will use the recruited tLogs to create a new `TagPartitionedLogSystem` for the new generation.
```

Comment:
Known committed version is the highest version that a proxy told a given TLog that it had durably committed on all TLogs.  The Master's is the maximum of all of that.  Known committed version is important, because it defines the lower bound of what range of mutations need to be copied to the new generation.

Reply:
Thanks for the native explanation! I shamelessly copied it into the doc.